### DISCUSSION TITLE
实用向的滤镜

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#120] [INFO 信息]
**欢迎补充更多案例！**

关联前置文档： [《音视频滤镜》](https://hooke007.github.io/unofficial/mpv_filters.html)

全部写进设置与快捷键显得啰嗦和冗长，大多数用户并不需要用到它们。
跳过赘述阶段，直接记录一些有实用意义的滤镜。使用精简语法，根据需求自行修改适配 input.conf / mpv.conf

>[!NOTE]
>本帖不记录 **d3d11vpp** 滤镜，它特定于 ``--gpu-context=d3d11`` 和具体硬件实现相关功能。当前的典型用例参考[《RTX VSR HDR》](https://github.com/hooke007/mpv_PlayKit/discussions/286)
>_本帖不记录 **vapoursynth系** 的滤镜，当前的可食用vs滤镜见wiki的 [相关页面](https://github.com/hooke007/mpv_PlayKit/wiki/2_portable_config#vs%E8%84%9A%E6%9C%AC%E8%B7%AF%E5%BE%84-)_

>[!NOTE]
>_此楼仅列举 **视频 vf** 及 **复合滤镜 lavfi-complex** 的示例，二楼为音频滤镜的示例。_
>点击以快速跳转至 👉 [**音频滤镜 af**](https://github.com/hooke007/mpv_PlayKit/discussions/120#discussioncomment-2724970) 部分

>[!TIP]
>一些滤镜有可选的着色器方案，它们通常是性能&解码兼容性更佳的上位替代品。

# 视频滤镜 --vf

因为大多数滤镜都只是软件实现没有硬件加速，所以确保在挂载前使用 `--hwdec=no` 或 `auto-copy` 以保证滤镜正常工作，但是对于有硬件解码支持的滤镜，应该切换到 `auto` / `yes` 或者 `auto-copy` 使其更高效执行。

### 硬解hack

通过前置一个滤镜 `format=fmt=yuv420p` （或其它受支持的格式） ，可以让绝大多数不支持原生硬解（ `--hwdec=auto` ）的滤镜强行支持，但是此时的整个处理链路的效率可能较低下，甚至可能不如直接使用 `--hwdec=auto-copy`
```
# mpv.conf
hwdec=d3d11va
vf=format=fmt=yuv420p,vflip,pad=aspect=16/12:color=Aqua

### 可自行比较处理效率

hwdec=d3d11va-copy
vf=vflip,pad=aspect=16/12:color=Aqua
```

## 格式转换

```
format=colorlevels=limited
format=primaries=bt.709
format=colormatrix=bt.709
format=gamma=gamma2.2
```
https://mpv.io/manual/master/#video-filters-format
重新标记色彩范围为有限，可用于修正一些片源被错误的标记为全范围（即full）
重新标记色彩原色为bt.709，可用于修正一些片源被错误标记（比如bt.470m）
重新标记色彩矩阵为bt.709，可用于修正一些片源被错误标记（比如smpte-240m）
修正伽马曲线为gamma2.2（显示器业界标准），用于补正mpv使用icc（bt.1886）时的亮度错误

- P.S. 不同于lavfi(ffmpeg)的 [同名滤镜](https://ffmpeg.org/ffmpeg-filters.html#format-1) mpv的 `format` 滤镜功能多且基本支持 **硬件加速** ；如果需要调用lavfi的同名滤镜，参考写法 `lavfi-format=pix_fmts=yuv444p16` 。

## 画面转动

```
vflip
hflip
```
垂直翻转
水平翻转

```
rotate=angle=90*PI/180
format=rotate=90
```
lavfi滤镜的顺时针90° https://ffmpeg.org/ffmpeg-filters.html#rotate
mpv原生滤镜的顺时针90°（仅90°/180°/270°时支持 **硬件加速** ）

- P.S.  它们不同于 https://mpv.io/manual/master/#options-video-rotate 的直接控制输出旋转（运行时修改会导致重载mpv进程）。在实际使用时，执行180°旋转时前两者都可以使用，90°/270°旋转推荐使用 `format=rotate` 或 `--video-rotate` （不会裁剪画面）。

>[!Tip]
>类似效果的着色器方案： 
>翻转 [flip_RT.glsl](../blob/main/portable_config/shaders/Canvas/flip_RT.glsl)
>旋转 [rotate_RT.glsl](../blob/main/portable_config/shaders/Canvas/rotate_RT.glsl)

## 黑边控制

```
pad=aspect=16/9:x=-1:y=-1
```
填充黑边至16:9并居中原帧画面，可用于2.35:1的片源适配16:9的字幕
https://ffmpeg.org/ffmpeg-filters.html#pad-1

```
crop=iw:ih-200:0:0,pad=iw:ih+200:0:200
```
切掉底部200像素行，填充顶部200像素行。实现视觉上的等效画面平移。（不使用原生的`--video-pan`的理由是这个参数影响字幕布局）
https://ffmpeg.org/ffmpeg-filters.html#crop

>[!Tip]
>类似效果的着色器方案：
>画面平移 [pan_RT.glsl](../blob/main/portable_config/shaders/Canvas/pan_RT.glsl)
>自适应黑边填充 [box_pad_ada.glsl](../blob/main/portable_config/shaders/Canvas/box_pad_ada.glsl)

## 畸变校正

```
lenscorrection=cx=0.5:cy=0.5:k1=-.25:k2=-.25
```
补正广角镜头的径向畸变。示例可用于GoPro
https://ffmpeg.org/ffmpeg-filters.html#lenscorrection

## 帧率变换

```
fps=fps=60/1.001
```
通过帧采样强制视频以指定帧率输出。示例可用于播放滚动弹幕时增强平滑性，此时分子数值（取整）为你的显示器刷新率
https://ffmpeg.org/ffmpeg-filters.html#fps-1
```
framerate=fps=60/1.001
```
作用类似，区别是帧混合（线性插值）。
https://ffmpeg.org/ffmpeg-all.html#framerate

- P.S. 此二项在效果上与补帧冲突。这亦会影响其它涉及帧率变化类选项的效果，例如 `--interpolation=yes` `--deterlance=yes`

```
mpdecimate=max=2
```
去除重复帧。特化案例下可用于补帧的前置滤镜。
https://ffmpeg.org/ffmpeg-filters.html#mpdecimate

## 尺寸调节

```
scale=w=-2:h=720
format=w=1280:h=720:convert=yes
```
lavfi的预缩放尺寸。示例为锁定高度720，宽度按源比例重设 https://ffmpeg.org/ffmpeg-filters.html#scale-1
mpv原生滤镜的对应功能较弱，不支持按比例只能手动指定双边数值，由于必须要使用 `convert=yes` 才生效，结果失去了硬件加速的支持。

- P.S. 如果需要固定尺寸播放这很有用，当然它只在窗口模式且不缩放窗口的情况下有效，可以考虑联合使用 https://mpv.io/manual/master/#options-video-unscaled 来实现真正的固定帧画面尺寸。

## 色彩控制

```
negate=components=y+u+v
lutyuv=y=negval:u=negval:v=negval

negate=components=r+g+b
lutrgb=r=negval:g=negval:b=negval
```
反相颜色（使图片具有底片质感）。两组分别适用于处理不同类型的源
negate滤镜可以精简输入为 `negate` 而无需补完后续的参数，它会自动识别源类型处理。
https://ffmpeg.org/ffmpeg-filters.html#negate
lutyuv/lutrgb中的 `negval` 等效 `maxval+minval-val`
https://ffmpeg.org/ffmpeg-filters.html#lut_002c-lutrgb_002c-lutyuv

- P.S. negate滤镜的缺点是处理YUV源时会引入细小的色彩偏移。
- P.S. lutyuv对YUV-ful_colorlevels源的动态处理存在问题，遇到这种情况需要在前面插入滤镜 `format=colorlevels=limited:convert=yes` 预转换片源为有限动态范围再交给lutyuv处理。

>[!Tip]
>等效的着色器方案：
>反相 [color_alt_luma.glsl](../blob/main/portable_config/shaders/color/color_alt_luma.glsl) [color_alt_rgb.glsl](../blob/main/portable_config/shaders/color/color_alt_rgb.glsl)

```
lutyuv=u=128:v=128
lutrgb=r=0:g=0
```
移除片源中指定的平面。两示例分别为仅保留LUMA（8bit）；仅保留BLUE通道

- P.S. lutyuv中要注意的问题同前；如果是10bit的片源，取值应该为512（即 (2^10)/2 ），如果面对更高位深时你不想做无谓的计算，可以使用参数 `(minval+maxval)/2` 作为值。

>[!Tip]
>等效的着色器方案：
>平面屏蔽 [plane_block_RT.glsl](../blob/main/portable_config/shaders/plane/plane_block_RT.glsl)

```
lut3d=file="使用绝对路径指定到你的3dlut文件"
```
导入3dlut进行调色
https://ffmpeg.org/ffmpeg-filters.html#lut3d-1

- P.S. 绝对路径中存在特殊字符且用在input.conf时，需要使用 **自定义引用** ，参考 [此处](https://hooke007.github.io/unofficial/mpv_filters.html#id7)。

>[!Important]
>该滤镜远不及 `--vo=gpu-next` 中的 [`--lut`](https://mpv.io/manual/master/#options-lut) [`--target-lut`](https://mpv.io/manual/master/#options-target-lut) [`--image-lut`](https://mpv.io/manual/master/#options-image-lut) 这些实现来的多变和高效。

```
colortemperature=temperature=6504
```
矫正色温到D65。我只知道11区的部分视频会以9300K拍摄出品，可以用此方法补正
https://ffmpeg.org/ffmpeg-filters.html#colortemperature

>[!Tip]
>类似效果的着色器方案：
>色温调节 [color_temperature_RT.glsl](../blob/main/portable_config/shaders/color/color_temperature_RT.glsl)

## 字幕渲染

```
subtitles=filename="使用绝对路径到目标视频/字幕文件"
```
独立于原生字幕渲染。它的一种用处在 https://github.com/hooke007/mpv_PlayKit/discussions/61 已记录，并以外部脚本的方式进行了局部实现。
https://ffmpeg.org/ffmpeg-filters.html#subtitles-1

# 复合滤镜 --lavfi-complex

https://mpv.io/manual/master/#options-lavfi-complex

## 同步播放

常见的一个用法是单个mpv进程中同时播放多个视频源。
由于该滤镜的特殊性，从命令行启动mpv比较方便（或者启动后手动导入多条视频轨）

示例一：双源水平并排
```powershell
./mpv "视频文件甲" --external-file="视频文件乙" --lavfi-complex="[vid1][vid2]hstack=inputs=2[vo]"
```

示例二：同源一分九阵列3x3
```
set lavfi-complex "[vid1]split=9[v1][v2][v3][v4][v5][v6][v7][v8][v9];[v1][v2][v3][v4][v5][v6][v7][v8][v9]xstack=inputs=9:layout=0_0|w0_0|w0+w1_0|0_h0|w0_h0|w0+w1_h0|0_h0+h1|w0_h0+h1|w0+w1_h0+h1[vo]"
```

恢复此滤镜需要刷新轨道：
```
set lavfi-complex "" ; set vid no ; set vid auto
```
或者使用 [input_plus.lua](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/scripts/input_plus.lua) 脚本中的命令
```
set lavfi-complex "" ; script-binding trackV_refresh
```

注意：
要求输入的源具有相同尺寸，不同尺寸的源应该联合用 `scale` 滤镜限制它。

如需更多源和不同排列方向则阅读下方的链接。
参考： 【 https://github.com/mpv-player/mpv/issues/3854 】 [【FFmpeg_wiki】](https://trac.ffmpeg.org/wiki/Create%20a%20mosaic%20out%20of%20several%20input%20videos%20using%20xstack)

**Comment by hooke007** [comment]
# 音频滤镜 --af

## 多声道下混
```
pan="stereo|FL=1.8*FC+0.7*FL+0.7*BL+0.5*LFE|FR=1.8*FC+0.7*FR+0.7*BR+0.5*LFE"
```
此方案作为 [FAQ人声增强](https://github.com/hooke007/mpv_PlayKit/wiki/0_FAQ#%E5%A4%9A%E5%A3%B0%E9%81%93%E9%9F%B3%E9%A2%91%E7%9A%84%E4%B8%8B%E6%B7%B7) 的额外方案，可通过增大其中的 `FC` 的系数来拉高对话的响度。
https://ffmpeg.org/ffmpeg-filters.html#pan-1

## 变速播放时修正音调

```
scaletempo=scale=1.0:stride=60:overlap=.20:search=14:speed=tempo
scaletempo2=min-speed=0.25:max-speed=8.0:search-interval=40:window-size=12
rubberband=transients=mixed:detector=compound:phase=laminar:window=standard:smoothing=off:formant=preserved:pitch=consistency:channels=together:engine=finer:pitch-scale=1.0
```
`scaletempo` 是旧版mpv变速播放时默认自动激活的音频滤镜。示例为默认参数。
其中的 stride 值增大可略提升性能，太高会导致在高速时跳音，在低速时出现回声；若值太低则会明显改变音调
https://mpv.io/manual/master/#audio-filters-scaletempo[

`scaletempo2` 是当前版本变速播放时默认自动激活的音频滤镜，示例为默认参数。
其中的 min/max-speed 在此速度的范围外直接静音处理
https://mpv.io/manual/master/#audio-filters-scaletempo2[

`rubberband` 另一个在变速时保持音调的内置滤镜，因可调参数多而略显复杂，详见官方文档，示例为默认参数。
https://mpv.io/manual/master/#audio-filters-rubberband

- P.S. 在mpv中使用时建议使用条件profile处理而不是常驻，示例：
```ini
[APC]
 profile-cond    = speed~=1
 af-append       = @apc:rubberband
[APC-alt]
 profile-cond    = speed==1
 af-remove       = @apc:rubberband
```

- P.S. 针对以上三个滤镜，mpv 内部做了一些限制和自动化处理，当设置 `audio-pitch-correction=no` 时无论插入哪个滤镜都无法生效，（默认值） `yes` 时自动插入 `scaletempo2` ，如果你设置了其它项比如 `rubberband` ，mpv将使用你指定的滤镜而不会挂载 `scaletempo2` 。

## 音频均衡器

对不同频段进行增益调节
```
superequalizer=1b=2.0:2b=3.6:3b=3.8:4b=5.5:5b=6.0:6b=6.4:7b=6.6:8b=6.4:9b=6.0:10b=5.2:11b=4.0:12b=3.2:13b=3.0:14b=3.2:15b=3.8:16b=4.4:17b=5.2:18b=6.5
equalizer
firequalizer
```
https://ffmpeg.org/ffmpeg-filters.html#superequalizer
https://ffmpeg.org/ffmpeg-filters.html#equalizer
https://ffmpeg.org/ffmpeg-filters.html#firequalizer

- P.S. 我只用 `superequalizer` 所以后两者没有给出可用子参数，但是有两更实用的脚本实现了运行时动态调节 [equalizer.lua](https://gist.github.com/avih/41acff712abd32e1f436235388c8b523) [firequalizer15.lua](https://github.com/mfcc64/mpv-scripts/blob/master/firequalizer15.lua) 。

# 复合滤镜 --lavfi-complex

## 多音轨同时播放

```
lavfi-complex="[aid1] [aid2] amix [ao]"
```
或者采用便捷的方式 [input_plus.lua](https://github.com/hooke007/mpv_PlayKit/blob/main/portable_config/scripts/input_plus.lua)
或者另一种非滤镜的实现 [audio-dupe.lua](https://gist.github.com/bitingsock/5e9714efff963c9689b0671d68f195ad)

**Comment by dyphire** [comment]
## 增加点音频滤镜相关内容
**双通道设备调节各通道音**
`loudnorm`和`dynnorm`音频过滤器用于调节各声道的增益。可解决双通道设备下多声道音频下混时产生的背景音或人声过小的问题
```
#loudnorm
af-toggle=@audnorm:lavfi=[loudnorm=I=-16:TP=-3:LRA=4] 
#dynaudnorm
af-toggle=@dynnorm:lavfi=[dynaudnorm=g=5:f=250:r=0.9:p=0.5]
```
另一组示例
```
af-toggle=@audnorm:!loudnorm=I=-25:TP=-1.5:LRA=1:linear=false
af-toggle=@dynnorm:!dynaudnorm=f=500:g=17:r=0.1
```
`loudnorm`为广播规范，通过EBU R.128进行响度标准化，使用最广
`dynnorm` 是“智能”响度归一化，不进行裁剪。动态地将规范化应用于文件。需谨慎使用，因为这可能会改变声音的特性

**dynnorm的动态调节菜单脚本：** [drcbox.lua](https://github.com/dyphire/mpv-scripts/blob/main/drcbox.lua)
修改自 https://gist.github.com/richardpl/0c8011dc23d7ac7b7831b2e6d680114f ，实现播放中使用`dynnorm`过滤器动态调整各通道增益
![Snipaste_2022-05-13_16-32-03](https://user-images.githubusercontent.com/41094733/168317443-b74614b7-8d06-4dbd-aaf6-33029aac74ba.jpg)

**修复mpv的音轨载入和切换时的兼容问题**
经测试发现切换高码率音轨、手动载入外部音轨时会导致视频出现长短不一的冻结卡顿的问题
相关issue：https://github.com/mpv-player/mpv/issues/9591
针对这一问题写了个脚本进行修复：[fix-avsync.lua](https://github.com/dyphire/mpv-scripts/blob/main/fix-avsync.lua)

**Reply by 906018080** [reply]
这个fix-avsync.lua脚本只是禁用音频滤镜吗? 
本来想要解决loudnorm滤镜导致的卡顿问题
结果使用脚本后发现loudnorm滤镜都不在工作状态

**Reply by dyphire** [reply]
~~是的，这脚本只是临时方案。真正解决要等mpv修复上面提到的issue。~~

这脚本一开始是为了解决切换音轨和载入外部音轨时出现的不明冻结卡顿现象，这方面的效果很好。
~~至于倍数播放相关的代码单纯是看到上面的issue后才塞进去的，效果不理想是预期内的（~~

edit: 音频滤镜卡顿相关的问题已被mpv上游于近期修复，相关临时代码已从fix-avsync.lua中删除

**Reply by hooke007** [reply]
建议：提问类开单独的帖子

**Reply by dyphire** [reply]
@yeezylife 你刚才那个问题比较简单，可以通过条件配置实现这一点。我直接贴一下吧，也算是该方案的补充
```
[audio-filter]
profile-desc=音频通道大于2时开启动态调节滤镜，适用于双通道设备
profile-cond=p["audio-params/channel-count"] > 2
profile-restore=copy-equal
af-pre=@dynaudnorm:lavfi=[dynaudnorm=g=5:f=250:r=0.9:p=0.5]
```

**Comment by kli2013** [comment]
https://github.com/kli2013/myrule/blob/master/mpv%E6%B7%BB%E5%8A%A0%E6%B7%BB%E5%8A%A0%E4%B8%89%E5%88%86%E5%B1%8F%E5%8F%B3%E9%94%AE%E8%8F%9C%E5%8D%95.bat

复合滤镜没有办法实时切换是吧?

**Reply by hooke007** [reply]
这涉及到具体怎么执行，没法统一讲能不能。比如音频复合和取消就毫无障碍。
我测试了视频的并列与取消，控制台执行多步指令是没问题的，但是只靠一个按键对应的命令，大概率不可能实现

说个无关的，你的播放方案不保险，很多互联网的视频本身就是多轨的，vid2被用来存储封面。

**Reply by kli2013** [reply]
好的,以后遇到问题我就明白是这个了
不过现在只是用来播放硬盘里的抖音视频,以前我都用FFmpeg转成成品,现在直接用滤镜,节约空间

**Reply by hooke007** [reply]
优化建议，减少2/3的解码开销（不用多次导入同一文件并行解码），而且不用担心其它轨道存在。
你的——
![Snipaste_2022-11-21_22-30-47](https://user-images.githubusercontent.com/41094733/203171270-8ad981ba-261e-445d-a8ac-15f1c863e193.jpg)
修改——
![Snipaste_2022-11-21_22-31-07](https://user-images.githubusercontent.com/41094733/203171273-f4d1d357-21ec-487a-852c-64b3d35f1142.jpg)

**Reply by kli2013** [reply]
谢谢分享,又多学一个命令

**Reply by kli2013** [reply]
恢复是[vid1]split=1[vo]吗,还是有其他更正确的命令

**Reply by hooke007** [reply]
`set lavfi-complex ""` split本身就是ffmpeg里的滤镜，恢复当然要移除它

**Reply by kli2013** [reply]
写入uosc右键菜单实时切换为""画面会卡住,只有声音,`控制台使用set lavfi-complex ""也一样
cycle-values lavfi-complex "[vid1]split=3[v1][v2][v3];[v1][v2]hstack[v_tmp];[v_tmp][v3]hstack[vo]" ""

**Reply by hooke007** [reply]
lavfi-complex 是个比较麻烦的滤镜，重置之后要强制刷新vid
```
set vid no ; set vid auto
```

**Reply by kli2013** [reply]
谢谢,搞定了
```
cycle-values lavfi-complex "[vid1]split=3[v1][v2][v3];[v1][v2]hstack[v_tmp];[v_tmp][v3]hstack[vo]" "" ;set vid no ; set vid auto
```

**Reply by hooke007** [reply]
又想起了这个帖子，我发现当时我很呆。之前处理的中途合并，完全产生了多余的的性能浪费。
```
[vid1]split=3[v1][v2][v3];[v1][v2][v3]hstack=inputs=3[vo]
```

**Reply by kli2013** [reply]
> 又想起了这个帖子，我发现当时我很呆。之前处理的中途合并，完全产生了多余的的性能浪费。
> 
> ```
> [vid1]split=3[v1][v2][v3];[v1][v2][v3]hstack=inputs=3[vo]
> ```

不懂,但是还是点赞更新

**Reply by hooke007** [reply]
这个场景已局部通过着色器方案实现。参见 [mirror_sp_RT](https://github.com/hooke007/mpv_PlayKit/wiki/4_GLSL#mirror-) 和 [它的参数](https://github.com/hooke007/mpv_PlayKit/discussions/410#discussioncomment-13159829)
